File: internal/registry/registry.go
package registry

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/newrelic/nr-diagnose/internal/schemas"
	"gopkg.in/yaml.v3"
)

// Agent holds all loaded data for a single agent definition.
type Agent struct {
	Manifest      schemas.AgentManifest
	InstallScript string
	Knowledge     AgentKnowledge
	Hints         schemas.DiagnosticHints
	RunbookIndex  schemas.RunbookIndex
	Dir           string
}

// AgentKnowledge holds the contents of the knowledge/ folder.
type AgentKnowledge struct {
	Prerequisites  string
	CommonFailures string
	References     string
}

// LoadAgent loads an agent definition from a directory.
func LoadAgent(agentDir string) (*Agent, error) {
	agent := &Agent{Dir: agentDir}

	manifestData, err := os.ReadFile(filepath.Join(agentDir, "manifest.yaml"))
	if err != nil {
		return nil, fmt.Errorf("reading manifest: %w", err)
	}
	if err := yaml.Unmarshal(manifestData, &agent.Manifest); err != nil {
		return nil, fmt.Errorf("parsing manifest: %w", err)
	}

	scriptData, err := os.ReadFile(filepath.Join(agentDir, "install.sh"))
	if err != nil {
		return nil, fmt.Errorf("reading install.sh: %w", err)
	}
	agent.InstallScript = string(scriptData)

	agent.Knowledge.Prerequisites = readFileOrEmpty(filepath.Join(agentDir, "knowledge", "prerequisites.md"))
	agent.Knowledge.CommonFailures = readFileOrEmpty(filepath.Join(agentDir, "knowledge", "common-failures.md"))
	agent.Knowledge.References = readFileOrEmpty(filepath.Join(agentDir, "knowledge", "references.md"))

	hintsData, err := os.ReadFile(filepath.Join(agentDir, "diagnostics", "hints.yaml"))
	if err == nil {
		yaml.Unmarshal(hintsData, &agent.Hints)
	}

	indexData, err := os.ReadFile(filepath.Join(agentDir, "runbook", "index.yaml"))
	if err == nil {
		yaml.Unmarshal(indexData, &agent.RunbookIndex)
	}

	return agent, nil
}

// ListAgents returns manifests of all agents in the given directory (excluding _template).
func ListAgents(agentsDir string) ([]schemas.AgentManifest, error) {
	entries, err := os.ReadDir(agentsDir)
	if err != nil {
		return nil, err
	}

	var agents []schemas.AgentManifest
	for _, entry := range entries {
		if !entry.IsDir() || strings.HasPrefix(entry.Name(), "_") {
			continue
		}
		manifestPath := filepath.Join(agentsDir, entry.Name(), "manifest.yaml")
		data, err := os.ReadFile(manifestPath)
		if err != nil {
			continue
		}
		var m schemas.AgentManifest
		if err := yaml.Unmarshal(data, &m); err != nil {
			continue
		}
		agents = append(agents, m)
	}
	return agents, nil
}

// FindAgentDir locates an agent directory by name within the agents folder.
func FindAgentDir(agentsDir, agentName string) (string, error) {
	dir := filepath.Join(agentsDir, agentName)
	if _, err := os.Stat(filepath.Join(dir, "manifest.yaml")); err != nil {
		return "", fmt.Errorf("agent %q not found in %s", agentName, agentsDir)
	}
	return dir, nil
}

func readFileOrEmpty(path string) string {
	data, err := os.ReadFile(path)
	if err != nil {
		return ""
	}
	return string(data)
}

File: nr_diagnose/registry.py
"""Agent registry - loads agent definitions from disk."""

import os
from pathlib import Path
from dataclasses import dataclass, field
from typing import List, Optional

import yaml

from .schemas import AgentManifest, DiagnosticHints, RunbookIndex, RunbookIndexEntry


@dataclass
class AgentKnowledge:
    prerequisites: str = ""
    common_failures: str = ""
    references: str = ""


@dataclass
class Agent:
    manifest: AgentManifest = field(default_factory=AgentManifest)
    install_script: str = ""
    knowledge: AgentKnowledge = field(default_factory=AgentKnowledge)
    hints: DiagnosticHints = field(default_factory=DiagnosticHints)
    runbook_index: RunbookIndex = field(default_factory=RunbookIndex)
    dir: str = ""


def load_agent(agent_dir: str) -> Agent:
    """Load an agent definition from a directory."""
    agent = Agent(dir=agent_dir)
    agent_path = Path(agent_dir)

    # Load manifest
    manifest_data = yaml.safe_load((agent_path / "manifest.yaml").read_text())
    agent.manifest = AgentManifest(
        name=manifest_data.get("name", ""),
        display_name=manifest_data.get("display_name", ""),
        description=manifest_data.get("description", ""),
        target_os=manifest_data.get("target_os", ""),
        ports=manifest_data.get("ports", []),
        services=manifest_data.get("services", []),
        prerequisites=manifest_data.get("prerequisites", []),
    )

    # Load install script
    agent.install_script = (agent_path / "install.sh").read_text()

    # Load knowledge files
    agent.knowledge = AgentKnowledge(
        prerequisites=_read_file_or_empty(agent_path / "knowledge" / "prerequisites.md"),
        common_failures=_read_file_or_empty(agent_path / "knowledge" / "common-failures.md"),
        references=_read_file_or_empty(agent_path / "knowledge" / "references.md"),
    )

    # Load diagnostic hints
    hints_path = agent_path / "diagnostics" / "hints.yaml"
    if hints_path.exists():
        hints_data = yaml.safe_load(hints_path.read_text()) or {}
        agent.hints = DiagnosticHints(
            priority_commands=hints_data.get("priority_commands", []),
            context_hints=hints_data.get("context_hints", []),
        )

    # Load runbook index
    index_path = agent_path / "runbook" / "index.yaml"
    if index_path.exists():
        index_data = yaml.safe_load(index_path.read_text()) or {}
        entries = [
            RunbookIndexEntry(pattern=e.get("pattern", ""), entry_file=e.get("entry_file", ""))
            for e in index_data.get("entries", [])
        ]
        agent.runbook_index = RunbookIndex(entries=entries)

    return agent


def list_agents(agents_dir: str) -> List[AgentManifest]:
    """Return manifests of all agents in the directory (excluding _template)."""
    agents: List[AgentManifest] = []
    agents_path = Path(agents_dir)

    if not agents_path.exists():
        return agents

    for entry in sorted(agents_path.iterdir()):
        if not entry.is_dir() or entry.name.startswith("_"):
            continue
        manifest_path = entry / "manifest.yaml"
        if not manifest_path.exists():
            continue
        try:
            data = yaml.safe_load(manifest_path.read_text())
            agents.append(AgentManifest(
                name=data.get("name", ""),
                display_name=data.get("display_name", ""),
                description=data.get("description", ""),
                target_os=data.get("target_os", ""),
                ports=data.get("ports", []),
                services=data.get("services", []),
                prerequisites=data.get("prerequisites", []),
            ))
        except Exception:
            continue

    return agents


def find_agent_dir(agents_dir: str, agent_name: str) -> str:
    """Locate an agent directory by name within the agents folder."""
    dir_path = Path(agents_dir) / agent_name
    if not (dir_path / "manifest.yaml").exists():
        raise FileNotFoundError(f'agent "{agent_name}" not found in {agents_dir}')
    return str(dir_path)


def _read_file_or_empty(path: Path) -> str:
    """Read a file or return empty string if it doesn't exist."""
    try:
        return path.read_text()
    except (FileNotFoundError, OSError):
        return ""
